GradeCenter.Services/Schools/SchoolService.cs

```diff
@@ -0,0 +1,117 @@
+﻿using System;
+using GradeCenter.Data.Models;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using GradeCenter.Data;
+using GradeCenter.Data.Models.Account;
+using System.Collections;
+
+namespace GradeCenter.Services.Schools
+{
+    public class SchoolService : ISchoolService
+    {
+        private readonly GradeCenterContext _db;
+        public SchoolService(GradeCenterContext gradeCenterContext)
+        {
+            _db = gradeCenterContext;
+        }
+
+        /// <summary>
+        /// Gets all existing school entries in the database.
+        /// </summary>
+        /// <returns></returns>
+        public IEnumerable<School> Read()
+        {
+            return _db.Schools.Where(school => school.IsActive)
+                .ToList();
+        }
+
+        /// <summary>
+        /// Takes an number as a paramater
+        /// and returns the associated 
+        /// School entity instance.
+        /// </summary>
+        /// <param name="number"></param>        
+        /// <returns></returns>
+        public School? GetSchoolByNumber(int number)
+        {
+            var school = _db?.Schools?.FirstOrDefault(school => school.Number == number);
+
+            return school;
+        }
+
+        /// <summary>
+        /// Creates a new School entity instance
+        /// in the database. Whilst ensuring there is a 
+        /// relationship with a principal entity.
+        /// </summary>
+        /// <param name="newSchool"></param>
+        /// <returns></returns> 
+        public async Task Create(School newSchool)
+        {
+            var school = new School
+            {
+                Name = newSchool.Name,
+                Number = newSchool.Number,
+                Address = newSchool.Address
+            };
+
+            await _db.Schools.AddAsync(school);
+
+            await _db.SaveChangesAsync();
+        }
+
+        /// <summary>
+        /// Updates an existing School 
+        /// entity instance in the database.
+        /// </summary>
+        /// <param name="number"></param>
+        /// <param name="updatedSchool"></param>
+        /// <param name="students"></param>
+        /// <param name="teachers"></param>
+        /// <param name="principal"></param>
+        /// <returns></returns>
+        public async Task Update(int number, School? updatedSchool, IEnumerable<User>? updatedStudents, IEnumerable<User>? updatedTeachers, User? updatedPrincipal)
```

Comment:
You should accept "School" object here as well as a parameter

Reply:
Personally, I don't see much use-case here on having two seperate objects as arguments. Instead:

-  The Id of the original school can be stored within the `updatedSchool` argument (This will come from the request model).
-  Leave the service layer to obtain the original School instance, instead of passing it.

Nevertheless, I can further embed an additional argument if this is considered - best practice.

Reply:
I think you misunderstood me. Let me try to explain - in one of the other comments I mentioned that that you need to change the School class.

- In School, add 2 lists - Students and Teachers
- In User class, remove the "Teacher" list and rename "Students" to children instead as its prolly confusing.
- Now in the Update method you to accept only 1 argument "School"

Reply:
Currently, adding two seperate collections to the School entity would not be possible, as there is already an established relationship with the `Principal`. This is solely the reason, I am gathering the collections from the Principal navigation property. 

Altering the both the User and School entities as follows:

**School:**
```
public class School
{
    //Omitted for brevity...
    public virtual User? Principal { get; set; }

    public ICollection<User> Students { get; set; }
    public ICollection<User> Teachers { get; set; }
}
```
**User:**
```
public class User : IdentityUser
{	//Omitted for brevity...
    [Required]
    [ForeignKey("School")]
    public string? SchoolId { get; set; }
    public virtual School School { get; set; }

    public ICollection<User> Children { get; set; } = new HashSet<User>();
    public ICollection<User> Parents { get; set; } = new HashSet<User>();
}
```
As well as having configured the following relationships within the DatabaseContext:
```
builder.Entity<School>()
             .HasMany(l => l.Students)
             .WithOne(l => l.School);

builder.Entity<School>()
             .HasMany(l => l.Teachers)
             .WithOne(l => l.School);
```
Will produce the following error when recreating the migration:
`Cannot create a relationship between 'School.Teachers' and 'User.School' because a relationship already exists between 'School.Students' and 'User.School'. Navigations can only participate in a single relationship. If you want to override an existing relationship call 'Ignore' on the navigation 'User.School' first in 'OnModelCreating'.`